File: nodejs/src/webhooks.ts
/**
 * Webhook Utilities
 *
 * Copyright (c) 2026 REDSCVRY TECHNOLOGY PRIVATE LIMITED
 * Licensed under the MIT License.
 */

import { createHmac, timingSafeEqual } from 'crypto';

export interface VerifySignatureParams {
  payload: string | Buffer;
  signature: string;
  secret: string;
}

/**
 * Verify a webhook signature.
 *
 * Structurify signs webhook payloads using HMAC-SHA256.
 * The signature is sent in the X-Structurify-Signature header.
 *
 * @example
 * ```typescript
 * import { verifyWebhookSignature } from '@structurify/sdk';
 *
 * app.post('/webhook', (req, res) => {
 *   const isValid = verifyWebhookSignature({
 *     payload: req.body,
 *     signature: req.headers['x-structurify-signature'],
 *     secret: 'your_webhook_secret',
 *   });
 *
 *   if (!isValid) {
 *     return res.status(401).json({ error: 'Invalid signature' });
 *   }
 *
 *   // Process webhook
 * });
 * ```
 */
export function verifyWebhookSignature(params: VerifySignatureParams): boolean {
  const { payload, signature, secret } = params;

  const payloadBuffer = Buffer.isBuffer(payload)
    ? payload
    : Buffer.from(payload, 'utf-8');

  // Remove "sha256=" prefix if present
  const sig = signature.startsWith('sha256=') ? signature.slice(7) : signature;

  const expected = createHmac('sha256', secret).update(payloadBuffer).digest('hex');

  try {
    return timingSafeEqual(Buffer.from(expected, 'hex'), Buffer.from(sig, 'hex'));
  } catch {
    return false;
  }
}

/**
 * Compute HMAC-SHA256 signature for a payload.
 * Useful for testing webhook handlers.
 */
export function computeSignature(payload: string | Buffer, secret: string): string {
  const payloadBuffer = Buffer.isBuffer(payload)
    ? payload
    : Buffer.from(payload, 'utf-8');

  const sig = createHmac('sha256', secret).update(payloadBuffer).digest('hex');
  return `sha256=${sig}`;
}

File: python/structurify/webhooks.py
"""
Webhook Utilities

Copyright (c) 2026 REDSCVRY TECHNOLOGY PRIVATE LIMITED
Licensed under the MIT License.
"""

import hashlib
import hmac
from typing import Union


def verify_signature(
    payload: Union[str, bytes],
    signature: str,
    secret: str,
) -> bool:
    """
    Verify a webhook signature.

    Structurify signs webhook payloads using HMAC-SHA256.
    The signature is sent in the X-Structurify-Signature header.

    Args:
        payload: The raw request body (string or bytes)
        signature: The signature from X-Structurify-Signature header
        secret: Your webhook secret

    Returns:
        True if signature is valid, False otherwise.

    Example:
        from structurify.webhooks import verify_signature

        @app.post("/webhook")
        def handle_webhook(request):
            payload = request.body
            signature = request.headers.get("X-Structurify-Signature")

            if not verify_signature(payload, signature, "your_secret"):
                return {"error": "Invalid signature"}, 401

            data = request.json()
            if data["event"] == "extraction.completed":
                # Handle extraction completion
                pass
    """
    if isinstance(payload, str):
        payload = payload.encode("utf-8")

    if isinstance(secret, str):
        secret = secret.encode("utf-8")

    # Remove "sha256=" prefix if present
    if signature.startswith("sha256="):
        signature = signature[7:]

    expected = hmac.new(secret, payload, hashlib.sha256).hexdigest()

    return hmac.compare_digest(expected, signature)


def compute_signature(payload: Union[str, bytes], secret: str) -> str:
    """
    Compute HMAC-SHA256 signature for a payload.

    This is primarily useful for testing webhook handlers.

    Args:
        payload: The payload to sign
        secret: Your webhook secret

    Returns:
        Signature string in "sha256=xxx" format.
    """
    if isinstance(payload, str):
        payload = payload.encode("utf-8")

    if isinstance(secret, str):
        secret = secret.encode("utf-8")

    sig = hmac.new(secret, payload, hashlib.sha256).hexdigest()
    return f"sha256={sig}"
